File: kiddytube_flutter/lib/parent/release_pin_policy.dart
import 'parent_pin.dart';

/// Rules for the factory development PIN (`2580`) and release gating.
/// Parity with Kotlin [ReleasePinPolicy].
class ReleasePinPolicy {
  ReleasePinPolicy._();

  /// When true, an entered PIN equal to the factory PIN must be rejected
  /// even if it still matches the stored hash.
  static bool rejectDefaultDevPin({
    required bool isDebugBuild,
    required bool releaseReady,
    required bool pinChangedFromDefault,
  }) {
    if (releaseReady) return true;
    if (!isDebugBuild && pinChangedFromDefault) return true;
    return false;
  }

  /// Release builds block channel/player launch until the factory PIN is replaced.
  static bool requirePinChangeForKidPlayback({
    required bool isDebugBuild,
    required bool pinChangedFromDefault,
  }) =>
      !isDebugBuild && !pinChangedFromDefault;

  static Future<bool> matchesDefaultDevPinAsync(String? salt, String? hash) async {
    if (salt == null || salt.isEmpty || hash == null || hash.isEmpty) {
      return false;
    }
    return ParentPinManager().verifyPinAsync(
      ParentPinManager.defaultDevPin,
      salt,
      hash,
    );
  }

  /// Sync helper for unit tests; prefer [sanitizePinFlagsAsync] on the UI path.
  static bool matchesDefaultDevPin(String? salt, String? hash) {
    if (salt == null || salt.isEmpty || hash == null || hash.isEmpty) {
      return false;
    }
    return ParentPinManager().verifyPin(
      ParentPinManager.defaultDevPin,
      salt,
      hash,
    );
  }

  static Future<({bool pinChangedFromDefault, bool releaseReady})>
      sanitizePinFlagsAsync({
    required String? pinSalt,
    required String? pinHash,
    required bool pinChangedFromDefault,
    required bool releaseReady,
  }) async {
    // Flags already say "still default" — skip PBKDF2 (cold-start / tests).
    if (!pinChangedFromDefault) {
      return (pinChangedFromDefault: false, releaseReady: false);
    }
    final stillDefault = await matchesDefaultDevPinAsync(pinSalt, pinHash);
    final changed = !stillDefault;
    final ready = releaseReady && changed;
    return (pinChangedFromDefault: changed, releaseReady: ready);
  }

  static ({bool pinChangedFromDefault, bool releaseReady}) sanitizePinFlags({
    required String? pinSalt,
    required String? pinHash,
    required bool pinChangedFromDefault,
    required bool releaseReady,
  }) {
    if (!pinChangedFromDefault) {
      return (pinChangedFromDefault: false, releaseReady: false);
    }
    final stillDefault = matchesDefaultDevPin(pinSalt, pinHash);
    final changed = !stillDefault;
    final ready = releaseReady && changed;
    return (pinChangedFromDefault: changed, releaseReady: ready);
  }
}

File: KiddyTube/app/src/main/java/ae/kiddytube/app/parent/ReleasePinPolicy.kt
package ae.kiddytube.app.parent

/**
 * Rules for the factory development PIN (`2580`) and release-ready gating.
 *
 * Debug builds keep the factory PIN for day-to-day development until Release ready is on.
 * Release builds allow the factory PIN only for first-run parent setup (until the PIN is
 * changed), then reject it. Kid playback on release builds requires a non-default PIN.
 */
object ReleasePinPolicy {
    /**
     * When true, an entered PIN equal to [ParentPinManager.DEFAULT_DEV_PIN] must be rejected
     * even if it still matches the stored hash.
     */
    fun rejectDefaultDevPin(
        isDebugBuild: Boolean,
        releaseReady: Boolean,
        pinChangedFromDefault: Boolean
    ): Boolean {
        if (releaseReady) return true
        // Release APK: after parent leaves the factory PIN, never accept 2580 again.
        if (!isDebugBuild && pinChangedFromDefault) return true
        return false
    }

    /** Release APKs block channel/player launch until the factory PIN is replaced. */
    fun requirePinChangeForKidPlayback(
        isDebugBuild: Boolean,
        pinChangedFromDefault: Boolean
    ): Boolean = !isDebugBuild && !pinChangedFromDefault

    /** True when the stored hash is still the factory development PIN. */
    fun matchesDefaultDevPin(salt: String?, hash: String?): Boolean {
        if (salt.isNullOrBlank() || hash.isNullOrBlank()) return false
        val expected = ParentPinManager.hashPin(ParentPinManager.DEFAULT_DEV_PIN, salt) ?: return false
        return expected.equals(hash, ignoreCase = true)
    }

    /**
     * Reconcile flags with the actual stored hash: default hash clears change/ready;
     * release-ready never sticks while still on the factory PIN.
     */
    fun sanitizePinFlags(
        pinSalt: String?,
        pinHash: String?,
        pinChangedFromDefault: Boolean,
        releaseReady: Boolean
    ): SanitizedPinFlags {
        val stillDefault = matchesDefaultDevPin(pinSalt, pinHash)
        val changed = pinChangedFromDefault && !stillDefault
        val ready = releaseReady && changed
        return SanitizedPinFlags(
            pinChangedFromDefault = changed,
            releaseReady = ready
        )
    }

    data class SanitizedPinFlags(
        val pinChangedFromDefault: Boolean,
        val releaseReady: Boolean
    )
}
